File: androidEats/src/main/java/com/dunk/eats/android/presentation/recipe_list/RecipeListViewModel.kt
package com.dunk.eats.android.presentation.recipe_list

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dunk.eats.domain.model.*
import com.dunk.eats.domain.util.ErrorMessageQueueUtil
import com.dunk.eats.domain.util.Queue
import com.dunk.eats.interactors.recipe_categories.FoodCategories
import com.dunk.eats.interactors.recipe_list.SearchRecipes
import com.dunk.eats.interactors.recipe_categories.CategoryTypes
import com.dunk.eats.presentation.recipe_list.RecipeListEvents
import com.dunk.eats.presentation.recipe_list.RecipeListState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.*
import javax.inject.Inject
import kotlin.collections.ArrayList

@HiltViewModel
class RecipeListViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val searchRecipes:SearchRecipes,
    private val categoryTypes: CategoryTypes
) : ViewModel() {

     val state: MutableState<RecipeListState> = mutableStateOf(RecipeListState())

    init {
       onTriggerEvent(RecipeListEvents.LoadRecipes )
    }

    fun onTriggerEvent(event:RecipeListEvents){
        when(event){
            RecipeListEvents.LoadRecipes -> {
                loadTopCategories()
                loadNewCategories()
                loadRecipes()
            }

            RecipeListEvents.NextPage -> {
                nextPage()
            }
            RecipeListEvents.NewSearch -> {
                newSearch()
            }
            is RecipeListEvents.OnUpdateQuery  -> {
                state.value = state.value.copy(query =  event.query, selectedFoodCategories = null)
            }
            is RecipeListEvents.OnSelectCategory  -> {
                onSelectCategory(event.foodCategories)
            }
            is RecipeListEvents.RemoveHeadMessageFromQueue -> {
                removeMessageAtHead()
            }
            else -> {
                addErrorToQueue(
                    ErrorMessage.Builder()
                        .id(UUID.randomUUID().toString())
                        .title("Error")
                        .uiComponentType(UIComponentType.Dialog)
                        .description("Invalid event")
                        .positive(PositiveAction(positiveBtnTxt = "Ok", onPositiveAction = {}))
                )
            }
        }
    }

    fun getFoodCategory(categoryName:String): FoodCategories? {
        return categoryTypes.getCategory(categoryName)
    }

    private fun onSelectCategory(foodCategories: FoodCategories) {
        state.value = state.value.copy(selectedFoodCategories = foodCategories, query = foodCategories.categoryName)
        newSearch()
    }

    private fun newSearch() {
        state.value = state.value.copy(page = 1, recipes = listOf() )
        loadRecipes()
    }

    private fun nextPage() {
        //update the state and increment the page
         state.value = state.value.copy(page = state.value.page + 1)
        loadRecipes()
    }

    private fun loadNewCategories(){
        state.value = state.value.copy(newCategories = categoryTypes.getNewCategories())
    }

    private fun loadTopCategories(){
        state.value = state.value.copy(topCategories = categoryTypes.getAllCategories().subList(0, 4))
    }

    private fun loadRecipes(){
        searchRecipes.execute(
            page = state.value.page,
            query = state.value.query,
        ).collectFlow(viewModelScope){ dataState ->
            state.value= state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { recipes ->
                appendRecipes(recipes)
            }

            dataState.errorMessage?.let { message ->
                addErrorToQueue(
                    ErrorMessage.Builder()
                        .id(UUID.randomUUID().toString())
                        .title(message.title)
                        .uiComponentType(UIComponentType.Dialog)
                        .description(message.description ?: "Unknown error")
                        .positive(PositiveAction(positiveBtnTxt = "Ok", onPositiveAction = {}))
                )
            }
        }
    }

    private fun appendRecipes(recipes: List<Recipe>){
        val curr = ArrayList(state.value.recipes)
        curr.addAll(recipes)
        state.value = state.value.copy(recipes = curr)
    }

    private fun addErrorToQueue(error: ErrorMessage.Builder) {
        if (!ErrorMessageQueueUtil().isErrorUnique(
                queue = state.value.errorQueue,
                message = error.build()
        )) {
            val queue = state.value.errorQueue
            queue.add(error.build())
            state.value = state.value.copy(errorQueue = queue)
        }
    }

    private fun removeMessageAtHead() {
        try {
            val queue = state.value.errorQueue
            queue.remove()
            // Forcing the app to recompose and remove the dialog since the queue is empty
            state.value = state.value.copy(errorQueue = Queue(mutableListOf()))
            state.value = state.value.copy(errorQueue = queue)
        }catch (e: Exception){
            // Nothing to remove queue is empty
        }
    }

}

File: iosEats/iosEats/presentation/recipe_list/RecipeListViewModel.swift
//
//  RecipeListViewModel.swift
//  iosEats
//
//  Created by Duncan K on 19/11/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI
import shared

class RecipeListViewModel: ObservableObject {
    
    let searchRecipes: SearchRecipes
    let foodCategories: CategoryTypes
    
    @Published var state: RecipeListState = RecipeListState()
    
    init(
        searchRecipes: SearchRecipes,
        foodCategories: CategoryTypes
    ){
        self.searchRecipes = searchRecipes
        self.foodCategories = foodCategories
        onTriggerEvent(stateEvent: RecipeListEvents.LoadRecipes())
    }
    
    func onTriggerEvent(stateEvent: RecipeListEvents){
        switch stateEvent {
        case is RecipeListEvents.LoadRecipes :
            loadRecipes()
        case is RecipeListEvents.NewSearch :
            doNothing()
        case is RecipeListEvents.NextPage :
            nextPage()
        case is RecipeListEvents.OnUpdateQuery :
            doNothing()
        case is RecipeListEvents.OnSelectCategory :
            doNothing()
        case is RecipeListEvents.RemoveHeadMessageFromQueue :
            doNothing()
        default:
            doNothing()
        }
    }
    
    private func nextPage(){
        let currentState = (self.state.copy() as! RecipeListState)
        updateState(page: Int(currentState.page + 1))
        loadRecipes()
    }
    
    private func loadRecipes() {
        let currentState = (self.state.copy() as! RecipeListState)
        do{
            try searchRecipes.execute(
                page: Int32(currentState.page),
                query: currentState.query
            ).collectFlow(coroutineScope: nil, callback: { dataState in
                if dataState != nil {
                    let data = dataState?.data
                    let message = dataState?.errorMessage
                    let loading = dataState?.isLoading ?? false
                    
                    self.updateState(isLoading: loading)
                    
                    if data != nil {
                        self.appendRecipes(recipes: data as! [Recipe])
                    }
                    if message != nil {
                        self.displayErrorMessageByUIComponentType(message)
                    }
                }
            })
        }catch{
            print("\(error)")
        }
    }
    
    func appendRecipes(recipes: [Recipe]){
        var currentState = (self.state.copy()) as! RecipeListState
        var currentRecipes = currentState.recipes
        currentRecipes.append(contentsOf: recipes)
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: currentState.page,
            query: currentState.query,
            recipes: currentRecipes,
            topCategories: currentState.topCategories,
            newCategories: currentState.newCategories,
            selectedFoodCategories: currentState.selectedFoodCategories,
            bottomRecipe: currentState.bottomRecipe,
            errorQueue: currentState.errorQueue
        )
        
        currentState = self.state.copy() as! RecipeListState
        self.updateBottomRecipe(recipe: currentState.recipes[currentState.recipes.count - 1])
        
    }
    
    func updateBottomRecipe(recipe:Recipe) {
        updateState(bottomRecipe: recipe)
    }
    
    func shouldLoadNextPage(recipe:Recipe) -> Bool {
        let currentState = (self.state.copy()) as! RecipeListState
        if (recipe.id == currentState.bottomRecipe?.id) {
            if (SearchRecipes.Companion().RECIPE_PAGINATION_PAGE_SIZE *
                currentState.page <= currentState.recipes.count){
                if(!currentState.isLoading){
                    return true
                }
            }
        }
        return false
    }
    
    func displayErrorMessageByUIComponentType( _ errorMessage: ErrorMessage?){
        //Todo handle errpr message
    }
    
    func doNothing(){
        
    }
    
    func updateState(
        isLoading: Bool? = nil,
        page:Int? = nil,
        query: String? = nil,
        bottomRecipe: Recipe? = nil,
        queue: Queue<ErrorMessage>? = nil
    ){
        let currentState = (self.state.copy() as! RecipeListState)
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            page: Int32(page ?? Int(currentState.page)),
            query: query ?? currentState.query,
            recipes: currentState.recipes,
            topCategories: currentState.topCategories,
            newCategories: currentState.newCategories,
            selectedFoodCategories: currentState.selectedFoodCategories,
            bottomRecipe: bottomRecipe ?? currentState.bottomRecipe,
            errorQueue: currentState.errorQueue
        )
    }
    
}
